### PR TITLE
JIT: Use `ValueTuple` argument for SIMD type lookup for `AdvSimd.Store*`

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Follow-up to #106765. `Store` and `StoreVectorAndZip`, like `StoreSelectedScalar`, have variants that operate on `ValueTuples` of `Vectors`, and in the case of the former intrinsic, the SIMD size has to be determined at runtime since it can operate on 64- and 128-bit `Vectors`. Thus, the JIT needs to determine the `Vector` size used by the intrinsic overloads' `ValueTuple` argument.

There are a few other intrinsics (like `LoadAndInsertScalar`) that also operate on `ValueTuples`, and the JIT currently differentiates between these variants with dedicated `NamedIntrinsics` (`LoadAndInsertScalarVector64x2`, `LoadAndInsertScalarVector64x3`, etc). I think these can be unified into one `NamedIntrinsic` and use the `ValueTuple` lookup logic, in case we want to clean these up.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
@tannergooding I did a quick parse of the intrinsics that take `ValueTuples`, and none of them need to look up their SIMD sizes at runtime, so are there any correctness issues we need to fix here? I only see potential for cleanup with this new functionality.

File: src/coreclr/jit/hwintrinsiclistarm64.h [hunk 1]
```diff
@@ -486,9 +486,9 @@ HARDWARE_INTRINSIC(AdvSimd,       ShiftRightLogicalScalar,
...
-HARDWARE_INTRINSIC(AdvSimd,       Store,                                                             8,      2,     {INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_invalid,        INS_invalid,        INS_st1_2regs,      INS_invalid},     HW_Category_MemoryStore,           HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialImport|HW_Flag_SpecialCodeGen|HW_Flag_NeedsConsecutiveRegisters)
+HARDWARE_INTRINSIC(AdvSimd,       Store,                                                             8,      2,     {INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_invalid,        INS_invalid,        INS_st1_2regs,      INS_invalid},     HW_Category_MemoryStore,           HW_Flag_BaseTypeFromSecondArg|HW_Flag_BaseTypeFromValueTupleArg|HW_Flag_SpecialImport|HW_Flag_SpecialCodeGen|HW_Flag_NeedsConsecutiveRegisters)
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsiclistarm64.h:489]: The SIMD size here isn't correct as the size is not always `8`. Consider, for example:
https://github.com/dotnet/runtime/blob/7ff684af49cdf008bebbe10efcc10a2649795d65/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.PlatformNotSupported.cs#L15019-L15024

or in the case of `AdvSimd_Arm64`:
https://github.com/dotnet/runtime/blob/7ff684af49cdf008bebbe10efcc10a2649795d65/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.PlatformNotSupported.cs#L3600-L3601

Comment (amanasifkhalid) [reply src/coreclr/jit/hwintrinsiclistarm64.h:489]: I see, so this one needs to look up the size at runtime, too. Let me fix that.

File: src/coreclr/jit/hwintrinsiclistarm64.h [hunk 2]
```diff
@@ -678,13 +678,13 @@ HARDWARE_INTRINSIC(AdvSimd_Arm64, ShiftRightArithmeticRoundedNarrowingSaturateUn
...
-HARDWARE_INTRINSIC(AdvSimd_Arm64, Store,                                                            16,      2,     {INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs},   HW_Category_MemoryStore,           HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialImport|HW_Flag_SpecialCodeGen|HW_Flag_NeedsConsecutiveRegisters)
+HARDWARE_INTRINSIC(AdvSimd_Arm64, Store,                                                            -1,      2,     {INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs,      INS_st1_2regs},   HW_Category_MemoryStore,           HW_Flag_BaseTypeFromSecondArg|HW_Flag_BaseTypeFromValueTupleArg|HW_Flag_SpecialImport|HW_Flag_SpecialCodeGen|HW_Flag_NeedsConsecutiveRegisters)
```
Comment (tannergooding) [inline src/coreclr/jit/hwintrinsiclistarm64.h:681]: I believe this one only has overloads dealing with Vector128

**Review by tannergooding — APPROVED** [review]
LGTM. I believe this is all the places that are using tuple and that had been changed from .NET 8, but I notably did not compare line by line to find any other places that may have gotten incorrectly changed

File: src/coreclr/jit/hwintrinsiclistarm64.h (hunk 2 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/hwintrinsiclistarm64.h]: You're right; fixed.

**Comment by amanasifkhalid** [comment]
> LGTM. I believe this is all the places that are using tuple and that had been changed from .NET 8, but I notably did not compare line by line to find any other places that may have gotten incorrectly changed

There is also `LoadAndInsertScalar`, but that one is special-cased such that during importation, we convert the `NamedIntrinsic` to one specifying the `ValueTuple` shape. `VectorTableLookup` also has a `ValueTuple` arg, but it doesn't have an immediate argument, and it has a special path in the importer anyway, so I don't think anything else needs to be fixed?

**Comment by amanasifkhalid** [comment]
@tannergooding should this be backported to .NET 9, since we fixed the SIMD size lookup for `AdvSimd.Store`? I see some logic during codegen that relies on the SIMD size being correct to pass the right parameters to the emitter.

**Pull request merged**